sql/sql_expression.cc

```diff
@@ -3609,6 +3610,50 @@ parse(ML::Parse_Context & context, bool allowUtf8) {
     return WhenExpression(result);
 }
 
+static bool filterWhen(ExpressionValue & val,
+                       const SqlRowScope & rowScope,
+                       const BoundSqlExpression & boundWhen)
+{
+    auto passValue = [&] (Date timestamp)
+        {
+            auto tupleScope
+                = SqlExpressionWhenScope
+                ::getRowScope(rowScope, timestamp);
+
+            return boundWhen(tupleScope, GET_LATEST).isTrue();
+        };
+
+    if (val.isEmbedding() || val.isAtom() || val.empty()) {
+        // Don't deconstruct an embedding or an atom; either pass or not
+        // based upon the timestamp (there is only a single timestamp)
+        return passValue(val.getEffectiveTimestamp());
+    }
+
+    StructValue kept;
+    kept.reserve(val.rowLength());
+
+    // TODO: we should do two passes to first mark and then extract
```

Comment:
I don't think this needs a ticket number, but it does merit a TODO as if ever it causes performance issues we have an idea how to fix it.